Foundation/NSDecimal.swift

```diff
@@ -1,155 +1,1838 @@
+
+        if !isNegative {
+            error = integerMultiply(&big,result,p10)
+        } else {
+            error = integerDivide(&big,result,p10)
+        }
+
+        if error != .noError && error != .lossOfPrecision {
+            return error;
+        }
+
+        for i in 0..<big._length {
+            result[i] = big[i]
+        }
+
+        result._length = big._length
+    }
+
+    var big = WideDecimal()
+    // Handle the rest of the power (<= maxpow10)
+    let p10 = WideDecimal(pow10[Int(power)])
+
+    if !isNegative {
+        error = integerMultiply(&big, result, p10)
+    } else {
+        error = integerDivide(&big, result, p10)
+    }
+
+    for i in 0..<big._length {
+        result[i] = big[i]
+    }
+
+    result._length = big._length
+
+    return error;
+}
+
+public func NSDecimalRound(_ result: UnsafeMutablePointer<Decimal>, _ number: UnsafePointer<Decimal>, _ scale: Int, _ roundingMode: NSDecimalNumber.RoundingMode) {
+    NSDecimalCopy(result,number) // this is unnecessary if they are the same address, but we can't test that here
+    result.pointee.round(scale: scale,roundingMode: roundingMode)
+}
 // Rounds num to the given scale using the given mode.
 // result may be a pointer to same space as num.
 // scale indicates number of significant digits after the decimal point
 
-public func NSDecimalNormalize(_ number1: UnsafeMutablePointer<Decimal>, _ number2: UnsafeMutablePointer<Decimal>, _ roundingMode: NSDecimalNumber.RoundingMode) -> NSDecimalNumber.CalculationError { NSUnimplemented() }
+public func NSDecimalNormalize(_ a: UnsafeMutablePointer<Decimal>, _ b: UnsafeMutablePointer<Decimal>, _ roundingMode: NSDecimalNumber.RoundingMode) -> NSDecimalNumber.CalculationError {
+    var diffexp = a.pointee.__exponent - b.pointee.__exponent
+    var result = WideDecimal()
+
+    //
+    // If the two numbers share the same exponents,
+    // the normalisation is already done
+    //
+    if diffexp == 0 {
+        return .noError
+    }
+
+    //
+    // Put the smallest of the two in aa
+    //
+    var aa: UnsafeMutablePointer<Decimal>
+    var bb: UnsafeMutablePointer<Decimal>
+
+    if diffexp < 0 {
+        aa = b
+        bb = a
+        diffexp = -diffexp
+    } else {
+        aa = a
+        bb = b
+    }
+
+    //
+    // Build a backup for aa
+    //
+    var backup = Decimal()
+
+    NSDecimalCopy(&backup,aa)
+
+    //
+    // Try to multiply aa to reach the same exponent level than bb
+    //
+
+    if integerMultiplyByPowerOf10(&result, WideDecimal(aa.pointee), Int(diffexp)) == .noError {
+        // Succeed. Adjust the length/exponent info
+        // and return no errorNSDecimalNormalize
+        aa.pointee._length = result._length
+        for i in 0..<result._length {
+            aa.pointee[i] = result[i]
+        }
+        aa.pointee._isCompact = 0
+        aa.pointee._exponent = bb.pointee._exponent
+        return .noError;
+    }
+
+    NSUnimplemented() // work in progress
```

Comment:
might be good to comment on the remaining work here that needs to be done as a reminder ;)